client/tinychain/ml/optimizer.py

```diff
@@ -27,13 +21,21 @@ def create(cls, learning_rate=F32(0.01)):
 
         return cls({"lr": learning_rate})
 
-    def optimize(self, i, gradients):
-        if isinstance(gradients, collections.abc.Mapping):
-            return {name: self.optimize(i, gradients[name]) for name in gradients}
-        elif isinstance(gradients, collections.abc.Iterable):
-            return [self.optimize(i, gradient) for gradient in gradients]
-        else:
-            return gradients * self["lr"]
+    def optimize(self, i, gradient, inputs, loss):
```

Comment:
Great suggestion! I think this could avoid double-computing the layer outputs also (currently they're computed once on the forward pass and once on the backward pass). Per offline conversaion, I'll merge in this PR and you can get started on refactoring `Differentiable` to provide a list of parameters for an `Optimizer` to optimize.